### PR TITLE
SDK updates for preview4

### LINKED ISSUE #263: Adapt NEP-5 to the new NEP-17
The [NEP-5](https://github.com/neo-project/proposals/blob/master/obsolete/nep-5.mediawiki) token standard was recently replaced by the [NEP-17](https://github.com/neo-project/proposals/pull/126) standard. We need to adapt the changes that affect neow3j and change the naming of NEP-5 related parts to NEP-17.

Check: https://github.com/neo-project/neo/pull/2024

### PR DESCRIPTION
**Pull request by mialbu**
Closes #263 #290 #291 #310 #312 #316 #328 #330 #332 #334

- [x] Adapt NEP-5 to NEP-17
- [x] Adapt RPC calls
- [x] Support Management contract and adapt contract deployment accordingly
- [x] Adapt opcode prices
- [x] Update native script hashes
- [x] Update integration tests
- [x] Update invocation result
- [x] Update manifest
- [x] Update application log
- [x] Adapt transaction fee calculation

### CODE REVIEW

File: utils/src/main/java/io/neow3j/contract/ScriptBuilder.java [hunk 1]
```diff
@@ -132,6 +132,12 @@ public ScriptBuilder pushParam(ContractParameter param) {
             case ARRAY:
                 pushArray((ContractParameter[]) value);
                 break;
+            case ANY:
+                if (value == null) {
+                    opCode(OpCode.PUSHNULL);
+                }
+                // TODO: 23.12.20 Michael: Serialize ContractParameter Any
```
Comment (csmuller) [inline utils/src/main/java/io/neow3j/contract/ScriptBuilder.java:139]: Create an issue for the missing serialization and deserialization of the Any contract parameter and remove this TODO.

File: contract/src/main/java/io/neow3j/contract/ManagementContract.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        }
+        int id = stackItem.asArray().get(0).asInteger().getValue().intValue();
+        int updateCounter = stackItem.asArray().get(1).asInteger().getValue().intValue();
+        String hash = Numeric.reverseHexString(stackItem.asArray().get(2).asByteString().getAsHexString());
+        String script = Numeric.toHexStringNoPrefix(stackItem.asArray().get(3).asByteString().getValue());
+        ContractManifest manifest = stackItem.asArray().get(4).asByteString().getAsJson(ContractManifest.class);
+        return new ContractState(id, updateCounter, hash, script, manifest);
+    }
+
+    // TODO: 23.12.20 Michael: add deploy, update and destroy methods
```
Comment (csmuller) [inline contract/src/main/java/io/neow3j/contract/ManagementContract.java:87]: As discussed, we only need the `deploy` method to be available here. `update` and `destroy` are only useful when called inside of a contract.

**Review by csmuller — CHANGES_REQUESTED** [review]

File: contract/src/main/java/io/neow3j/contract/PolicyContract.java [hunk 3]
```diff
@@ -82,27 +75,39 @@ public BigInteger getMaxBlockSystemFee() throws IOException {
...
+    }
 
-        return arrayItem.asArray().getValue()
-                .stream().map(ac -> ScriptHash.fromAddress(ac.asByteString().getAsAddress()))
-                .collect(Collectors.toList());
+    /**
+     * Checks whether an account is blocked in the Neo network.
+     *
+     * @param scriptHash the script hash of the account.
+     * @return whether the account is blocked or not.
```
[the remark this answers is not in the document]
Comment (mialbu) [reply contract/src/main/java/io/neow3j/contract/PolicyContract.java:106]: added

File: core/src/main/java/io/neow3j/transaction/WitnessScope.java [hunk 4]
```diff
@@ -9,9 +9,10 @@
 public enum WitnessScope {
 
     /**
-     * This scope is used to mark the transaction sender.
+     * This scope is used only to mark the transaction sender.
+     * No contract is witnessed.
```
[the remark this answers is not in the document]
Comment (mialbu) [reply core/src/main/java/io/neow3j/transaction/WitnessScope.java:13]: I updated the javadoc 👍🏼

File: contract/src/main/java/io/neow3j/contract/TransactionBuilder.java [hunk 5]
```diff
@@ -318,9 +326,20 @@ private long getSystemFeeForScript() throws IOException {
         // check in the smart contract.
         Signer[] signers = this.signers.toArray(new Signer[0]);
         String script = Numeric.toHexStringNoPrefix(this.script);
-        NeoInvokeScript response = neow.invokeScript(script, signers).send();
-        // The GAS amount is returned in fractions (10^8)
-        return Long.parseLong(response.getInvocationResult().getGasConsumed());
+        NeoInvokeScript response = neow.invokeScript(Base64.encode(Numeric.hexStringToByteArray(script)), signers).send();
+        // TODO: 23.12.20 Michael: check whether the returned system fee already includes
```
[the remark this answers is not in the document]
Comment (mialbu) [reply contract/src/main/java/io/neow3j/contract/TransactionBuilder.java:330]: No check is needed, the returned `gasconsumed` value from the `invokescript` rpc call represents the actual system fee in gas that will be consumed by invoking that script.
I removed the todo.

File: core/src/main/java/io/neow3j/protocol/core/JsonRpc2_0Neow3j.java [hunk 6]
```diff
@@ -321,6 +334,8 @@ public JsonRpc2_0Neow3j(
 
     // SmartContract Methods
 
+    // TODO: 22.12.20 Michael: invokecontractverify
```
[the remark this answers is not in the document]
Comment (mialbu) [reply core/src/main/java/io/neow3j/protocol/core/JsonRpc2_0Neow3j.java:337]: Yes, it was merged in [this](https://github.com/neo-project/neo-modules/pull/409) pr. I'll create an issue for it.

File: core/src/main/java/io/neow3j/protocol/core/JsonRpc2_0Neow3j.java [hunk 7]
```diff
@@ -451,6 +466,15 @@ public JsonRpc2_0Neow3j(
                 NeoImportPrivKey.class);
     }
 
+    @Override
+    public Request<?, NeoCalculateNetworkFee> calculateNetworkFee(String tx) {
```
[the remark this answers is not in the document]
Comment (mialbu) [reply core/src/main/java/io/neow3j/protocol/core/JsonRpc2_0Neow3j.java:470]: Good input! We could use the description of the calls directly from [here](https://docs.neo.org/v3/docs/en-us/reference/rpc/latest-version/api.html) in the official Neo docs.

[the remark this answers is not in the document]
Comment (mialbu) [reply core/src/main/java/io/neow3j/protocol/core/JsonRpc2_0Neow3j.java:470]: I changed `tx` to `txHex`.

File: contract/src/main/java/io/neow3j/contract/ManagementContract.java (hunk 2 above)
Comment (mialbu) [reply contract/src/main/java/io/neow3j/contract/ManagementContract.java:87]: I removed the update and destroy from the todo and also the according unneeded constants.

File: core/src/main/java/io/neow3j/protocol/core/JsonRpc2_0Neow3j.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mialbu) [reply core/src/main/java/io/neow3j/protocol/core/JsonRpc2_0Neow3j.java]: I've created an issue for this (#343).

File: utils/src/main/java/io/neow3j/contract/ScriptBuilder.java (hunk 1 above)
Comment (mialbu) [reply utils/src/main/java/io/neow3j/contract/ScriptBuilder.java]: I created an issue (#344) for that.

File: contract/src/main/java/io/neow3j/contract/PolicyContract.java [hunk 8]
```diff
@@ -82,27 +75,39 @@ public BigInteger getMaxBlockSystemFee() throws IOException {
...
-                .getInvocationResult().getStack().get(0);
+    public BigInteger getExecFeeFactor() throws IOException {
+        return callFuncReturningInt(GET_EXEC_FEE_FACTOR);
+    }
 
-        if (!arrayItem.getType().equals(StackItemType.ARRAY)) {
-            throw new UnexpectedReturnTypeException(arrayItem.getType(), StackItemType.ARRAY);
-        }
+    /**
+     * Gets the storage price.
```
[the remark this answers is not in the document]
Comment (mialbu) [reply contract/src/main/java/io/neow3j/contract/PolicyContract.java:93]: The storage price was added to the native policy contract in [this](https://github.com/neo-project/neo/pull/2045/files) pr. I couldn't find a detailed description whether it represents the price per byte or similar. Maybe you are more successful in finding a detailed explanation for this one, @csmuller.

File: contract/src/main/java/io/neow3j/contract/TransactionBuilder.java [hunk 9]
```diff
@@ -290,7 +287,18 @@ Transaction buildTransaction() throws Throwable {
...
-        long networkFee = calcNetworkFee() + additionalNetworkFee;
+
+        Transaction tx = new Transaction(neow, version, nonce, validUntilBlock, signers, systemFee,
+                0, attributes, script, new ArrayList<>());
+
+        byte[] txBytes = tx.getHashData();
+        getSignerAccounts().forEach(signerAcc -> {
+            byte[] verifScript = signerAcc.getVerificationScript().getScript();
+            tx.addWitness(new Witness(txBytes, verifScript));
+        });
```
[the remark this answers is not in the document]
Comment (mialbu) [reply contract/src/main/java/io/neow3j/contract/TransactionBuilder.java]: Agree. I moved the technical code to the private method `calcNetworkFee` as you suggested. 👍🏼

File: contract/src/test/java/io/neow3j/contract/SmartContractTest.java [hunk 10]
```diff
@@ -196,88 +199,76 @@ public void invokeShouldProduceCorrectScript() throws Throwable {
...
         setUpWireMockForCall("invokefunction", "invokescript_registercandidate.json",
-                NEO_SCRIPT_HASH.toString(), NEP5_TOTALSUPPLY);
-        SmartContract sc = new SmartContract(NEO_SCRIPT_HASH, this.neow);
+                NEO_SCRIPT_HASH.toString(), NEP17_TOTALSUPPLY);
+        SmartContract sc = new SmartContract(NEO_SCRIPT_HASH, neow);
         expectedException.expect(UnexpectedReturnTypeException.class);
         expectedException.expectMessage(new StringContains(StackItemType.BOOLEAN.jsonValue()));
-        sc.callFuncReturningInt(NEP5_TOTALSUPPLY);
+        sc.callFuncReturningInt(NEP17_TOTALSUPPLY);
     }
```
[the remark this answers is not in the document]
Comment (mialbu) [reply contract/src/test/java/io/neow3j/contract/SmartContractTest.java:193]: I've just added test for this method.

**Review by csmuller — APPROVED** [review]

**Comment by gsmachado** [comment]
@csmuller the build problems are not related to this PR? If yes, then we could merge it to move on. What do you think?

**Comment by csmuller** [comment]
> @csmuller the build problems are not related to this PR? If yes, then we could merge it to move on. What do you think?

They are related to this PR. It should not be merged yet.

**Review by gsmachado — APPROVED** [review]

**Comment by gsmachado** [comment]
@mialbu @csmuller 

It's funny: even if we merged this PR, the issues did not close. Anyway, please, double-check if issues are closed once we merge stuff, please. 😄

**Pull request merged**